fun.lua

```diff
@@ -1032,6 +1040,86 @@ methods.operator = operator
 exports.op = operator
 methods.op = operator
 
+--------------------------------------------------------------------------------
+-- Stateful iterators
+--------------------------------------------------------------------------------
+
+-- Returns 'gen' throwing an error that explains that method is not supported
+local unsupported_stateful_method_gen = function(method)
+    return function()
+        error(string.format("stateful iterators do not support method '%s'",
+              method))
+    end
+end
+
+-- Helper of 'wrap_with_elems'
+local gen_with_elems_wrapper = function(gen_param_elems, state)
+    local elems = gen_param_elems[3]
+    if not elems then
+        return gen_param_elems[1](gen_param_elems[2], state)
+    else
+        gen_param_elems[3] = nil
+        return state, unpack(elems)
+    end
+end
+
+-- Wrap an iterator that yields all values from 'elems' table on the first
+-- iteration and then works in the same way as 'wrap(gen, param, state)'
+local wrap_with_elems = function(elems, gen, param, state)
+    local gen_param_elems = {gen, param, elems}
+    return wrap(gen_with_elems_wrapper, gen_param_elems, state)
+end
+
+-- Checks if drop_while should continue skipping. If iterator is not exhausted,
+-- elements returned by iterator are wrapped into a table and returned as the
+-- third return value
+local stateful_drop_while_check = function(fun, state_x, ...)
+    if state_x == nil then
+        return state_x, false
+    end
+    return state_x, fun(...), {...}
+end
+
+-- Implementation of method drop_while: original one copies iterator state to
+-- iterate over the same element twice - it doesn't work with stateful iterators
+local stateful_drop_while = function(fun, gen_x, param_x, state_x)
+    assert(type(fun) == "function", "invalid first argument to drop_while")
+    local cont = true
+    local elems
+    while state_x ~= nil and cont do
+        state_x, cont, elems = stateful_drop_while_check(fun, gen_x(param_x, state_x))
+    end
+    if state_x == nil then
+        return wrap(nil_gen, nil, nil)
+    end
+    return wrap_with_elems(elems, gen_x, param_x, state_x)
+end
+
+local stateful_methods = deepcopy(methods)
```

Comment:
I'm curious: should an iterator created from a stateful iterator be stateful?

For example, should the following expression work correctly?

```lua
<stateful iterator>:filter():drop_while()
```

If the answer is 'yes', whether it works this way (seems no) and whether it is complicated to implement?

I guess we can store the `wrap` implementation in the metatable and use it in the methods like `:filter()` instead of just `wrap`, which produces a stateless (usual) luafun iterator.

Reply:
Outdated: stateful iterators were dropped.